### PR TITLE
Add GetFieldDomainNames method to retrieve all field domain names which are available for a dataset

### PR DESCRIPTION
**Pull request by nyalldawson**
~~Implemented for OpenFileGDB driver only for now~~

~~The FileGDB and GPKG drivers will require special implementations of this, which will be done shortly. For now I'm after general feedback on the API.~~

Implemented for all drivers which currently support field domains

### CODE REVIEW

File: gcore/gdaldataset.cpp [hunk 1]
```diff
@@ -8437,6 +8437,58 @@ void GDALDatasetClearStatistics(GDALDatasetH hDS)
...
+
+/** Returns a list of the names of all field domains stored in the dataset.
+ *
+ * @param papszOptions Driver specific options determining how attributes
+ * should be retrieved. Pass nullptr for default behavior.
+ *
+ * @return list of field domain names
+ * @since GDAL 3.5
+ */
+std::vector<std::string> GDALDataset::GetFieldDomainNames(CSLConstList) const{
```
Comment (rouault) [inline gcore/gdaldataset.cpp:8453]: to make Doxygen happy
```suggestion
std::vector<std::string> GDALDataset::GetFieldDomainNames(CPL_UNUSED CSLConstList papszOptions) const{
```

File: gcore/gdaldataset.cpp [hunk 2]
```diff
@@ -8437,6 +8437,58 @@ void GDALDatasetClearStatistics(GDALDatasetH hDS)
     GDALDataset::FromHandle(hDS)->ClearStatistics();
 }
 
+
+/************************************************************************/
+/*                        GetFieldDomains()                              */
```
Comment (rouault) [inline gcore/gdaldataset.cpp:8442]:
```suggestion
/*                      GetFieldDomainNames()                            */
```

**Review by rouault — COMMENTED** [review]
A note in the Doxygen comments of GDALDataset::AddFieldDomain() and GDALDataset::GetFieldDomain() that drivers should make sure to update  m_oMapFieldDomains for the default implementation of GetFieldDomainNames() to
work, otherwise they need to make a specialized implementation of it, could be helpful

**Comment by rouault** [comment]
rebasing on master will fix the failure in the build-mac cmake CI target (due to https://github.com/OSGeo/gdal/pull/5156 being applied)

**Comment by nyalldawson** [comment]
@rouault I've now implemented for all drivers which support field domains

File: ogr/ogrsf_frmts/gpkg/ogrgeopackagedatasource.cpp [hunk 3]
```diff
@@ -7176,6 +7176,50 @@ const char* GDALGeoPackageDataset::GetGeometryTypeString(OGRwkbGeometryType eTyp
...
+        CPLError(CE_Warning, CPLE_AppDefined,
+                 "Number of rows returned for field domain names has been "
+                 "truncated.");
+    }
+    else if( oResultTable->RowCount() > 0 )
+    {
+        oDomainNamesList.reserve(oResultTable->RowCount());
+        for ( int i = 0; i < oResultTable->RowCount(); i++ )
+        {
+            const char *pszConstraintName = oResultTable->GetValue(0, i);
```
Comment (rouault) [inline ogr/ogrsf_frmts/gpkg/ogrgeopackagedatasource.cpp:7215]: pszConstraintName should be tested against nullptr (even if the official schema from the GPKG spec would have a not null constraint, nothing guarantee that the database we open respect it)

Comment (nyalldawson) [reply ogr/ogrsf_frmts/gpkg/ogrgeopackagedatasource.cpp:7215]: Fixed, thanks

**Comment by nyalldawson** [comment]
Thanks @rouault !

**Pull request merged**